pilot/pkg/model/jwks_resolver.go

```diff
@@ -126,7 +175,95 @@ func (r *jwksURIResolver) resolveJwksURIUsingOpenID(issuer string) (string, erro
 	}
 
 	// Set JwksUri in cache.
-	r.cache.Set(issuer, jwksURI)
+	r.JwksURICache.Set(issuer, jwksURI)
 
 	return jwksURI, nil
 }
+
+func (r *jwksResolver) getRemoteContent(url string) ([]byte, error) {
+	resp, err := r.client.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		_ = resp.Body.Close()
+	}()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return body, nil
+}
+
+func (r *jwksResolver) refresher() {
+	// Wake up once in a while and refresh stale items.
+	r.refreshTicker = time.NewTicker(r.refreshInterval)
+	for {
+		select {
+		case now := <-r.refreshTicker.C:
+			r.refresh(now)
+		case <-r.closing:
+			r.refreshTicker.Stop()
+			return
+		}
+	}
+}
+
+func (r *jwksResolver) refresh(t time.Time) {
+	var wg sync.WaitGroup
+	hasChange := false
+
+	r.keyEntries.Range(func(key interface{}, value interface{}) bool {
+		jwksURI := key.(string)
+
+		e := value.(*jwtPubKeyEntry)
+		oldPubKey := e.pubKey
+
+		// key rotation: fetch JWT public key again if it's expired.
+		if e.expireTime.Before(t) {
```

Comment:
Is it OK that is the entry is expired, there could potentially be a large number of go routines created, all to fetch an update to the same key?

Reply:
Thanks for the feedback Martin ! For now I added a LastUsedTime property in cachedItem, and updated refresher job to evict cached items which haven't been used for a while. We may need some eviction mechanism like LRU to prevent cache growing indefinitely in the future.